### ISSUE TITLE
`test foo =~ bar` causes crash

### ISSUE THREAD

**Issue by McDutchie** [att/ast#1152]
**Description of problem:**

Using the `=~` operator with the `test`/`[` builtin reliably causes ksh to crash.

**Ksh version:**

Version A 2017.0.0-devel-2327-gb56dd2a

**Steps to reproduce:**

    $ ../ksh -c 'test foo =~ foo'
    Abort trap: 6
    $ ../ksh -c '[ foo =~ foo ]'
    Abort trap: 6

**Comment by krader1961** [comment]
Somewhat surprisingly this is a case where the failure does not occur with ksh93u+ or ksh93v-; at least not on macOS and Linux. So thanks for the bug report, @McDutchie. It looks like we've broken something. Whomever fixes this needs to ensure the reproducer gets added as a regression unit test.

**Comment by krader1961** [comment]
Note that the failure is because we're calling `abort()` from `test_binop()` because argument `op` isn't a recognized value. Something that would be caught at compile time if `op` was an enum rather than a simple int.

**Comment by krader1961** [comment]
Oh, and, ksh93v- simply does this if the `op` isn't recognized:

```
/* NOTREACHED */
return(0);
```

Replacing the `return` with `abort()` causes the exact same failure we're seeing on the master branch. That is, the `op` value is not recognized and the `/* NOTREACHED */` comment is horse shit. So this is not a regression introduced by the work done over the past year+.

**Comment by krader1961** [comment]
Note that bash reports an error:

```
$ bash -c 'test foo =~ foo'
bash: line 0: test: =~: binary operator expected
```

While zsh complains about `~ not found`. You've gotta love POSIX shell compliance. 😸

**Comment by krader1961** [comment]
P.P.S., I'm going to fix this by ensuring that all valid `op` values are explicitly handled and if any unrecognized `op` value is received we still abort. That will match the old behavior and correctly handle any case where an invalid `op` (that doesn't happen to look like a valid op) is received.

**Comment by krader1961** [comment]
Sigh. Given how `test` operators are defined (e.g., `TEST_NE`) it won't be (easily) possible to convert the values to a simple enum that would ensure mistakes like this are caught by the compiler or lint tools. This is exhibit #65536 about the original authors of the code being too clever by half.

**Comment by krader1961** [comment]
I don't think we want to change the behavior found in ksh93u+. Specifically, that `test foo =~ foo` is always false with an exit status of one and no error message. I agree that is broken but to maintain backward compatibility I'm going to reinstate that behavior.

**Comment by krader1961** [comment]
Gah! I've changed my mind after writing the text for a new issue reminding us to change this in a future release. Since this is a syntax error, and sets status to non-zero regardless of how we handle it, we might as well bite the bullet now and emit a syntax error and set status to two.

**Comment by McDutchie** [comment]
Op 09-02-19 om 03:48 schreef Kurtis Rader:
> Closed #1152 <https://github.com/att/ast/issues/1152> via 0481d50 
> <https://github.com/att/ast/commit/0481d509c52cba3a913ef82daea007892500a515>.

Shouldn't "operator not support" be "operator not supported"?

- M.

-- 
modernish -- harness the shell
https://github.com/modernish/modernish

**Comment by siteshwar** [comment]
> Op 09-02-19 om 03:48 schreef Kurtis Rader:
> Closed #1152 <#1152> via [0481d50](https://github.com/att/ast/commit/0481d509c52cba3a913ef82daea007892500a515) <[0481d50](https://github.com/att/ast/commit/0481d509c52cba3a913ef82daea007892500a515)>.
> Shouldn't "operator not support" be "operator not supported"? - M.
> […](#)
> -- modernish -- harness the shell https://github.com/modernish/modernish

I noticed it today and fixed it already https://github.com/att/ast/commit/a127eeb7bae890b9c56c8376f59f15da1c0e098a.
